piff/readers.py

```diff
@@ -0,0 +1,273 @@
+                   entries may be `None`, and if the WCS `dict` is `None` the
+                   pointing coord is always `None`.
+        """
+        raise NotImplementedError()
+
+    def nested(self, name):
+        """Return a context manager that yields a new `Reader` that reads
+        content written by a similarly nested `Writer`.
+
+        :param name:     Base name for all objects read with the returned object.
+
+        :returns:     A context manager that yields a nested reader object.
+        """
+        raise NotImplementedError()
+
+    def get_full_name(self, name):
+        """Return the full name of a data structure saved with the given name,
+        combining it with any base names added if this `Reader` was created by
+        the `nested` method.
+        """
+        raise NotImplementedError()
+
+
+class FitsReader(Reader):
+    """A `Reader` implementation that reads from to multiple FITS HDUs.
+
+    This reader is intended to read files written by Piff before the `Reader`
+    and `Writer` abstractions were added.
+
+    `FitsReader.nested` yields a reader that reads from the same file and
+    prepends all names with its base name to form the EXTNAME, concatenating
+    them with ``_``.
+
+    :param fits:        An already-open `fitsio.FITS` object.
+    :param base_name:   Base name to prepend to all object names, or `None`.
+    """
+    def __init__(self, fits, base_name):
+        self._fits = fits
+        self._base_name = base_name
+
+    @classmethod
+    @contextmanager
+    def _open(cls, file_name):
+        """Return a context manager that opens the given FITS file and yields
+        a `FitsReader` object.
+
+        :param file_name:   Name of the file to read.
+
+        :returns:  A context manager that yields a `FitsReader` instance.
+        """
+        with fitsio.FITS(file_name, "r") as f:
+            yield cls(f, base_name=None)
+
+    def read_struct(self, name):
+        # Docstring inherited.
+        extname = self.get_full_name(name)
+        if extname not in self._fits:
+            return None
+        cols = self._fits[extname].get_colnames()
+        data = self._fits[extname].read()
+        assert len(data) == 1
+        struct = dict([ (col, data[col][0]) for col in cols ])
+        for k, v in struct.items():
+            # Convert numpy scalar types to native Python types.  We assume all
+            # bytes are supposed to be strs in read_struct, but not in
+            # read_table.  Conversions from numeric types are important because
+            # numpy scalars don't always satisfy the expected isinstance
+            # checks.
+            if isinstance(v, bytes):
+                struct[k] = v.decode()
+            elif isinstance(v, np.bool_):
+                struct[k] = bool(v)
+            elif isinstance(v, np.integer):
+                struct[k] = int(v)
+            elif isinstance(v, np.floating):
+                struct[k] = float(v)
```

Comment:
This whole bit wasn't in the original, and maybe isn't necessary?  We just returned struct in `read_kwargs`.  So, (a) can you explain a little more why you think this loop is useful? And (b) if it is important to keep, add unit tests for the bytes and bool branches that are currently uncovered.  (We apparently don't have cases of those right now.)

Reply:
I've been burned in the past by numpy scalar types not being exactly substitutable for their built-in counterparts, and have gotten in the habit of coding defensively around them.  IIRC that's the only rationale for all but the `bytes`, case, which was necessary to have `str` round-trip through the FITS binary table.  But let me see what happens if I drop some or all of these and get back to you with a more precise answer.

Reply:
Ok, I've confirmed that you're right: none of this is necessary, not even the `bytes` -> `str` coercion, and removed it all.  No idea what got me confused on that point.